### PR TITLE
Fix escaping nested close brackets

### LINKED ISSUE #3193: using {{}} in config
If I understood everything correctly, then ..

https://github.com/NLog/NLog/wiki/AppDomain-Layout-Renderer

This does not work
```
${appdomain:format={1} - {0}}
```
This works
```
${appdomain:format={1\\}}
```

This does not work
```
="...${appdomain:format={1\\} - {0\\}}"
```

In sum, these functions do not give the desired result.
ParseParameterName - Removes single slashes, with double does not work correctly
ParseParameterValue - It does not work correctly without a slash

### PR DESCRIPTION
**Pull request by lobster2012-user**
fixes #3193

### CODE REVIEW

File: tests/NLog.UnitTests/Layouts/LayoutParserTests.cs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        [InlineData(@"                           ${cached:${appdomain:format={0\} {1\}}}")]
+        [InlineData(@"                  ${cached:${cached:${appdomain:format={0\} {1\}}}}")]
+        [InlineData(@"         ${cached:${cached:${cached:${appdomain:format={0\} {1\}}}}}")]
+        [InlineData(@"${cached:${cached:${cached:${cached:${appdomain:format={0\} {1\}}}}}}")]
+        [InlineData(@"                                    ${literal:text={0\} {1\}}")]
+        [InlineData(@"                           ${cached:${literal:text={0\} {1\}}}")]
+        [InlineData(@"                  ${cached:${cached:${literal:text={0\} {1\}}}}")]
+        [InlineData(@"         ${cached:${cached:${cached:${literal:text={0\} {1\}}}}}")]
+        [InlineData(@"${cached:${cached:${cached:${cached:${literal:text={0\} {1\}}}}}}")]
+        public void Issue_3193_Nested_Сlosing_Braces(string input)
```
Comment (304NotModified) [inline tests/NLog.UnitTests/Layouts/LayoutParserTests.cs]: Hi! Thanks for the included tests!

Unfortunately it's a bit hard to follow. We are trying to use AAA in unit tests,see https://www.c-sharpcorner.com/UploadFile/dacca2/fundamental-of-unit-testing-understand-aaa-in-unit-testing/

File: tests/NLog.UnitTests/Layouts/LayoutParserTests.cs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                Assert.Equal("{0} {1}", textRenderer.Text);
+            }
+            else if (input.IndexOf("appdomain") >= 0)
+            {
+                var appDomain = LogFactory.CurrentAppDomain;
+                Assert.Equal($"{appDomain.Id} {appDomain.FriendlyName}", textRenderer.Text);
+            }
+            else
+            {
+                throw new Xunit.Sdk.XunitException("NOT SUPPORTED");
```
Comment (304NotModified) [inline tests/NLog.UnitTests/Layouts/LayoutParserTests.cs]: Is this assert.Fail?

**Comment by 304NotModified** [comment]
Any idea if this will fix also #2844

**Comment by lobster2012-user** [comment]
The question author needed to use double backslashes. I would not say that there is a problem at all. If he had more nested renders, then yes, my pr would fix his problem.

File: tests/NLog.UnitTests/Layouts/LayoutParserTests.cs (hunk 2 above)
Comment (lobster2012-user) [reply tests/NLog.UnitTests/Layouts/LayoutParserTests.cs]: For NUnit - yes
```NUnit.Framework.Assert.Fail("..")```
https://stackoverflow.com/questions/14631923/cannot-find-assert-fail-and-assert-pass-or-equivalent

I updated test.
I think, whether to move the code in ```SimpleLayoutTests.cs```.

**Comment by pablinos** [comment]
Oooo, this is exciting, I'll have to try this out. I never properly solved the problem I was having, but I notice a few other items referencing the issue #2844 and as @lobster2012-user can't reproduce it, it may well be that it's been fixed. 

I'll have another try and let you know.

**Comment by lobster2012-user** [comment]
(I moved comment)
I updated test.
I think, whether to move the code in ```SimpleLayoutTests.cs```.

**Comment by 304NotModified** [comment]
> I think, whether to move the code in `SimpleLayoutTests.cs`.

I think that's better, as the test is "acting" on SimpleLayout.

File: tests/NLog.UnitTests/Layouts/LayoutParserTests.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        [Theory]
+        [InlineData(@"                                    ${literal:text={0\} {1\}}")]
+        [InlineData(@"                           ${cached:${literal:text={0\} {1\}}}")]
+        [InlineData(@"                  ${cached:${cached:${literal:text={0\} {1\}}}}")]
+        [InlineData(@"         ${cached:${cached:${cached:${literal:text={0\} {1\}}}}}")]
+        [InlineData(@"${cached:${cached:${cached:${cached:${literal:text={0\} {1\}}}}}}")]
+        public void Issue_3193_Nested_Сlosing_Braces(string input)
+        {
+            SimpleLayout simple = input.Trim();
+            var result = simple.Render(new LogEventInfo { });
```
Comment (304NotModified) [inline tests/NLog.UnitTests/Layouts/LayoutParserTests.cs]: Please ` LogEventInfo.CreateNullEvent()` :) (that's consistent with other tests)

File: tests/NLog.UnitTests/Layouts/LayoutParserTests.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        /// https://github.com/NLog/NLog/issues/3193
+        /// </summary>
+        /// <param name="input"></param>
+        [Theory]
+        [InlineData(@"                                    ${literal:text={0\} {1\}}")]
+        [InlineData(@"                           ${cached:${literal:text={0\} {1\}}}")]
+        [InlineData(@"                  ${cached:${cached:${literal:text={0\} {1\}}}}")]
+        [InlineData(@"         ${cached:${cached:${cached:${literal:text={0\} {1\}}}}}")]
+        [InlineData(@"${cached:${cached:${cached:${cached:${literal:text={0\} {1\}}}}}}")]
+        public void Issue_3193_Nested_Сlosing_Braces(string input)
```
Comment (304NotModified) [inline tests/NLog.UnitTests/Layouts/LayoutParserTests.cs]: Proposal for test name: Render_EscapedBrackets_ShouldRenderAllBrackets

(using scheme: MethodName_StateUnderTest_ExpectedBehavior)
